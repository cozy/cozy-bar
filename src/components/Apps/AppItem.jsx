import React from 'react'
import { appShape } from 'proptypes/index'

import AppIcon from 'cozy-ui/react/AppIcon'
import AppLinker from 'cozy-ui/react/AppLinker'
import HomeIcon from 'components/Apps/IconCozyHome'
import { translate } from 'cozy-ui/react/I18n'
import stack from 'lib/stack'
import PropTypes from 'prop-types'

export class AppItem extends React.Component {
  constructor(props) {
    super(props)
    this.onAppSwitch = this.onAppSwitch.bind(this)
  }

  componentWillUnmount() {
    if (this.switchTimeout) clearTimeout(this.switchTimeout)
  }

  onAppSwitch() {
    const { onAppSwitch } = this.props
    if (typeof onAppSwitch === 'function') {
      this.switchTimeout = setTimeout(() => {
        onAppSwitch()
      }, 1000)
    }
  }

  render() {
    const { t, useHomeIcon, app } = this.props

    const dataIcon = app.slug ? `icon-${app.slug}` : ''

    return (
      <AppLinker onAppSwitch={this.onAppSwitch} slug={app.slug} href={app.href}>
        {({ onClick, href }) => {
          const label = t(`${app.slug}.name`, {
            _: app.namePrefix ? `${app.namePrefix} ${app.name}` : app.name
          })
          return (
            <li
              className={`coz-nav-apps-item${
                app.isCurrentApp ? ' --current' : ''
              }`}
            >
              <a
                role="menuitem"
                href={href}
                data-icon={dataIcon}
                title={label}
                onClick={onClick}
              >
                {useHomeIcon ? (
                  <HomeIcon className="coz-nav-apps-item-icon" />
                ) : (
                  <AppIcon
                    app={app}
                    className="coz-nav-apps-item-icon"
                    key={app.slug}
                    {...stack.get.iconProps()}
                  />
                )}
                <p className="coz-label">{label}</p>
              </a>
            </li>
          )
        }}
      </AppLinker>
    )
  }
}

AppItem.propTypes = {
  app: appShape.isRequired,
  useHomeIcon: PropTypes.bool
}

export default translate()(AppItem)
