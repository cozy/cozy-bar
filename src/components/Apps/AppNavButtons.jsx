import React, { Component } from 'react'
import { connect } from 'react-redux'

import { getHomeApp } from 'lib/reducers'

import { translate } from 'cozy-ui/react/I18n'
import Icon from 'cozy-ui/react/Icon'
import HomeIcon from 'components/Apps/IconCozyHome'
import { isFetchingApps } from 'lib/reducers'

class AppNavButton extends Component {
  render() {
    const {
      homeApp,
      handleClick,
      appName,
      appNamePrefix,
      appSlug,
      iconPath,
      isFetchingApps,
      isPublic,
      opened,
      t
    } = this.props

    const isHomeApp = homeApp && homeApp.isCurrentApp

    if (!isPublic && isFetchingApps) {
      return (
        <div className="coz-nav-apps-btns --loading">
          <div className="coz-nav-apps-btns-home coz-loading-placeholder" />
          <div className="coz-nav-apps-btns-main coz-loading-placeholder" />
        </div>
      )
    }

    const displayName =
      !isHomeApp && appNamePrefix
        ? [
            t(`${appSlug}.name_prefix`, {
              _: appNamePrefix
            }),
            t(`${appSlug}.name`, {
              _: appName
            })
          ].join(' ')
        : t(`${appSlug}.name`, { _: appName })

    const homeHref = !isPublic && homeApp && homeApp.href

    return (
      <div className={`coz-nav-apps-btns${isHomeApp ? ' --currentHome' : ''}`}>
        {homeHref ? (
          <a href={homeHref} className="coz-nav-apps-btns-home">
            <HomeIcon className="coz-nav-apps-btns-home-svg" />
          </a>
        ) : (
          <span className="coz-nav-apps-btns-home">
            <HomeIcon className="coz-nav-apps-btns-home-svg" />
          </span>
        )}
        {!isHomeApp && <span className="coz-nav-apps-btns-sep" />}
        <button
          type="button"
          onClick={isPublic ? null : handleClick}
          className="coz-nav-apps-btns-main"
          aria-controls="coz-nav-pop--apps"
          data-tutorial="apps"
          disabled={isPublic}
        >
          {!isHomeApp && (
            <img className="coz-bar-hide-sm" src={iconPath} width="28" alt="" />
          )}
          <span className="coz-nav-app-name">{displayName}</span>
          {!isPublic && (
            <Icon icon={opened ? 'top' : 'bottom'} color="#95999d" size="12" />
          )}
        </button>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  homeApp: getHomeApp(state),
  isFetchingApps: isFetchingApps(state)
})

const mapDispatchToProps = () => ({})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AppNavButton))
