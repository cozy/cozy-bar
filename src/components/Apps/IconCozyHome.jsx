import React, { PureComponent } from 'react'

import AppIcon from 'cozy-ui/react/AppIcon'

import stack from 'lib/stack'
import homeIcon from 'assets/sprites/icon-cozy-home.svg'

class IconCozyHome extends PureComponent {
  render() {
    const { className } = this.props
    const fetchIcon = () =>
      `${stack.get.cozyURL()}/assets/images/icon-cozy-home.svg`

    return (
      <AppIcon
        fetchIcon={fetchIcon}
        fallbackIcon={homeIcon}
        className={className}
      />
    )
  }
}

export default IconCozyHome
