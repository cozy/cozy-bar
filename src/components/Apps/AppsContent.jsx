import React, { Component } from 'react'
import { connect } from 'react-redux'
import Proptypes from 'prop-types'

import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import { getApps, getHomeApp, isFetchingApps } from 'lib/reducers'

import AppItem from 'components/Apps/AppItem'
import AppItemPlaceholder from 'components/Apps/AppItemPlaceholder'
import cozyIcon from 'assets/icons/16/icon-cozy-16.svg'

const sorter = fn => (itemA, itemB) => fn(itemA) > fn(itemB)

export class AppsContent extends Component {
  constructor(props, context) {
    super(props, context)
    this.translateApp = translateApp(this.props.t)
  }

  render() {
    const {
      t,
      apps,
      breakpoints,
      homeApp,
      isFetchingApps,
      onAppSwitch
    } = this.props
    const { isMobile } = breakpoints
    const isHomeApp = homeApp && homeApp.isCurrentApp
    const homeSlug = homeApp && homeApp.slug

    if (!isFetchingApps && (!apps || !apps.length)) {
      return <p className="coz-nav--error coz-nav-group">{t('no_apps')}</p>
    }

    return (
      <div className="coz-nav-pop-content">
        <ul className="coz-nav-group">
          {isMobile && homeApp && (
            <AppItem app={homeApp} useHomeIcon onAppSwitch={onAppSwitch} />
          )}
          {isFetchingApps
            ? new Array(3)
                .fill({})
                .map((nothing, index) => <AppItemPlaceholder key={index} />)
            : apps
                .filter(app => app.slug !== homeSlug)
                .sort(sorter(this.translateApp))
                .map((app, index) => (
                  <AppItem app={app} key={index} onAppSwitch={onAppSwitch} />
                ))}
        </ul>
        {homeApp && !isMobile && !isHomeApp && (
          <a role="menuitem" href={homeApp.href} className="coz-apps-home-btn">
            <img src={cozyIcon} />
            {t('menu.home')}
          </a>
        )}
      </div>
    )
  }
}
AppsContent.propTypes = {
  homeApp: Proptypes.shape({
    isCurrentApp: Proptypes.bool,
    slug: Proptypes.string
  }),
  apps: Proptypes.array,
  isFetchingApps: Proptypes.bool.isRequired
}
const translateApp = t => app => {
  const namePrefix = app.namePrefix
    ? t(`${app.slug}.namePrefix`, { _: app.namePrefix })
    : null
  const name = t(`${app.slug}.name`, { _: app.name })
  return namePrefix ? `${namePrefix} ${name}` : `${name}`
}

const mapStateToProps = state => ({
  apps: getApps(state),
  homeApp: getHomeApp(state),
  isFetchingApps: isFetchingApps(state)
})

export default connect(mapStateToProps)(
  translate()(withBreakpoints()(AppsContent))
)
