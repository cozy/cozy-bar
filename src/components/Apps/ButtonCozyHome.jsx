import React from 'react'
import PropTypes from 'prop-types'

import flag from 'cozy-flags'
import { isFlagshipApp } from 'cozy-device-helper'
import { useWebviewIntent } from 'cozy-intent'

import IconCozyHome from './IconCozyHome'

export const ButtonCozyHome = ({ homeHref, isInvertedTheme }) => {
  const webviewIntent = useWebviewIntent()

  if (isFlagshipApp() || flag('flagship.debug'))
    return (
      <a
        onClick={() => {
          webviewIntent.call('backToHome')
        }}
        className="coz-nav-apps-btns-home coz-nav-apps-btns-home--is-flagship"
        data-testid="buttonCozyHome"
      >
        <IconCozyHome
          className="coz-nav-apps-btns-home-svg"
          isInvertedTheme={isInvertedTheme}
        />
      </a>
    )

  if (homeHref) {
    return (
      <a href={homeHref} className="coz-nav-apps-btns-home">
        <IconCozyHome
          className="coz-nav-apps-btns-home-svg"
          isInvertedTheme={isInvertedTheme}
        />
      </a>
    )
  }

  return (
    <span className="coz-nav-apps-btns-home">
      <IconCozyHome
        className="coz-nav-apps-btns-home-svg"
        isInvertedTheme={isInvertedTheme}
      />
    </span>
  )
}

ButtonCozyHome.propTypes = {
  homeHref: PropTypes.string,
  isInvertedTheme: PropTypes.bool
}
