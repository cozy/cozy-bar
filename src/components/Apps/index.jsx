import React, { Component } from 'react'
import PropTypes from 'prop-types'

import AppsContent from 'components/Apps/AppsContent'
import AppNavButtons from 'components/Apps/AppNavButtons'

class Apps extends Component {
  constructor(props) {
    super(props)
    this.state = {
      opened: false
    }
  }

  componentDidMount() {
    document.body.addEventListener('click', this.onClickOutside)
    this.modalContainer = document.getElementById('cozy-bar-modal-dom-place')
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.onClickOutside)
  }

  onClickOutside = event => {
    if (this.state.opened) {
      // if it's not a cozy-bar nav popup, close the opened popup
      if (
        !this.rootRef.contains(event.target) &&
        !this.modalContainer.contains(event.target)
      ) {
        this.setState({ opened: false })
        event.stopPropagation()
      }
    }
  }

  toggleMenu = () => {
    this.setState({ opened: !this.state.opened })
  }

  // data-tutorial attribute allows to be targeted in an application tutorial
  render() {
    const {
      appName,
      appNamePrefix,
      appSlug,
      iconPath,
      isPublic,
      isInvertedTheme
    } = this.props
    const { opened } = this.state
    return (
      <nav
        className="coz-nav coz-nav-apps"
        ref={ref => {
          this.rootRef = ref
        }}
      >
        <AppNavButtons
          appName={appName}
          appNamePrefix={appNamePrefix}
          appSlug={appSlug}
          iconPath={iconPath}
          handleClick={this.toggleMenu}
          opened={opened}
          isPublic={isPublic}
          isInvertedTheme={isInvertedTheme}
        />
        <div
          className="coz-nav-pop coz-nav-pop--apps"
          id="coz-nav-pop--apps"
          aria-hidden={!opened}
        >
          <AppsContent isInvertedTheme={isInvertedTheme} />
        </div>
      </nav>
    )
  }
}

Apps.propTypes = {
  appName: PropTypes.string,
  appNamePrefix: PropTypes.string,
  appSlug: PropTypes.string,
  iconPath: PropTypes.string,
  isPublic: PropTypes.bool,
  isInvertedTheme: PropTypes.bool
}

export default Apps
