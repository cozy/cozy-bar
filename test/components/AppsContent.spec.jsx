import React from 'react'
import { AppsContent } from 'components/Apps/AppsContent'
import { shallow } from 'enzyme'
import { tMock } from '../jestLib/I18n'

describe('Apps Content', () => {
  it('should render no App', () => {
    const root = shallow(
      <AppsContent
        t={tMock}
        breakpoints={{ isMobile: true }}
        isFetchingApps={false}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })
  it('should render display Drive', () => {
    const apps = [
      {
        slug: 'cozy-drive',
        name: 'Drive'
      }
    ]
    const root = shallow(
      <AppsContent
        t={tMock}
        apps={apps}
        breakpoints={{ isMobile: true }}
        isFetchingApps={false}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should render Loader when fetching', () => {
    const apps = [
      {
        slug: 'cozy-drive',
        name: 'Drive'
      }
    ]
    const root = shallow(
      <AppsContent
        t={tMock}
        apps={apps}
        breakpoints={{ isMobile: true }}
        isFetchingApps={true}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should display home in the app items lists, if we are on mobile', () => {
    const apps = [
      {
        slug: 'cozy-home',
        name: 'Home'
      }
    ]
    const root = shallow(
      <AppsContent
        t={tMock}
        apps={apps}
        homeApp={{
          slug: 'cozy-home'
        }}
        breakpoints={{ isMobile: true }}
        isFetchingApps={false}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should not display home within the app items list, but just a link', () => {
    const apps = [
      {
        slug: 'cozy-home',
        name: 'Home'
      }
    ]
    const root = shallow(
      <AppsContent
        t={tMock}
        apps={apps}
        homeApp={{
          slug: 'cozy-home'
        }}
        breakpoints={{ isMobile: false }}
        isFetchingApps={false}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should not display home if not on mobile and home is the currentApp', () => {
    const apps = [
      {
        slug: 'cozy-home',
        name: 'Home'
      }
    ]
    const root = shallow(
      <AppsContent
        t={tMock}
        apps={apps}
        homeApp={{
          slug: 'cozy-home',
          isCurrentApp: true
        }}
        breakpoints={{ isMobile: false }}
        isFetchingApps={false}
      />
    )
    expect(root.getElement()).toMatchSnapshot()
  })
})
