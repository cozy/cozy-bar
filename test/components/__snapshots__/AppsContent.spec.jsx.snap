// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apps Content should display home in the app items lists, if we are on mobile 1`] = `
<div
  className="coz-nav-pop-content"
>
  <ul
    className="coz-nav-group"
  >
    <Wrapper
      app={
        Object {
          "slug": "cozy-home",
        }
      }
      useHomeIcon={true}
    />
  </ul>
</div>
`;

exports[`Apps Content should not display home if not on mobile and home is the currentApp 1`] = `
<div
  className="coz-nav-pop-content"
>
  <ul
    className="coz-nav-group"
  />
</div>
`;

exports[`Apps Content should not display home within the app items list, but just a link 1`] = `
<div
  className="coz-nav-pop-content"
>
  <ul
    className="coz-nav-group"
  />
  <a
    className="coz-apps-home-btn"
    role="menuitem"
  >
    <img
      src="test-file-stub"
    />
    Back to home
  </a>
</div>
`;

exports[`Apps Content should render Loader when fetching 1`] = `
<div
  className="coz-nav-pop-content"
>
  <ul
    className="coz-nav-group"
  >
    <AppItemPlaceholder />
    <AppItemPlaceholder />
    <AppItemPlaceholder />
  </ul>
</div>
`;

exports[`Apps Content should render display Drive 1`] = `
<div
  className="coz-nav-pop-content"
>
  <ul
    className="coz-nav-group"
  >
    <Wrapper
      app={
        Object {
          "name": "Drive",
          "slug": "cozy-drive",
        }
      }
    />
  </ul>
</div>
`;

exports[`Apps Content should render no App 1`] = `
<p
  className="coz-nav--error coz-nav-group"
>
  No applications found on the Cozy.
</p>
`;
